janusgraph-core/src/main/java/org/janusgraph/graphdb/util/SubqueryIterator.java

```diff
@@ -17,80 +17,200 @@
 
 package org.janusgraph.graphdb.util;
 
-import java.util.ArrayList;
-import java.util.Iterator;
-import java.util.List;
-import java.util.function.Function;
-import java.util.stream.Stream;
-
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
 import org.janusgraph.core.JanusGraphElement;
 import org.janusgraph.core.JanusGraphException;
 import org.janusgraph.diskstorage.BackendTransaction;
 import org.janusgraph.graphdb.database.IndexSerializer;
+import org.janusgraph.graphdb.query.Query;
 import org.janusgraph.graphdb.query.graph.JointIndexQuery;
 import org.janusgraph.graphdb.query.profile.QueryProfiler;
 
-import com.google.common.cache.Cache;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 /**
  * @author davidclement90@laposte.net
  */
 public class SubqueryIterator implements Iterator<JanusGraphElement>, AutoCloseable {
 
-    private final JointIndexQuery.Subquery subQuery;
-
     private final Cache<JointIndexQuery.Subquery, List<Object>> indexCache;
 
     private Iterator<? extends JanusGraphElement> elementIterator;
 
     private List<Object> currentIds;
 
-    private QueryProfiler profiler;
+    private QueryProfiler streamedQueryProfiler;
+
+    private JointIndexQuery.Subquery streamedQuery;
 
     private boolean isTimerRunning;
 
-    public SubqueryIterator(JointIndexQuery.Subquery subQuery, IndexSerializer indexSerializer, BackendTransaction tx,
-            Cache<JointIndexQuery.Subquery, List<Object>> indexCache, int limit,
-            Function<Object, ? extends JanusGraphElement> function, List<Object> otherResults) {
-        this.subQuery = subQuery;
+    public SubqueryIterator(List<JointIndexQuery.Subquery> queries, IndexSerializer indexSerializer,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The constructor 'SubqueryIterator(List, IndexSerializer, BackendTransaction, Cache, int, Function)' has an NPath complexity of 8333](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5179944)